File: impl/go/internal/tool/websearch.go
// websearch — web search with a configurable provider (Go parity of websearch.ts).
// Provider order: an API key (Tavily/Exa) if configured, else a keyless scrape of
// DuckDuckGo's HTML endpoint. The scrape path means search works out of the box
// without a key, at the cost of lower reliability (rate-limits, layout drift) — so
// it fails soft rather than crashing the loop. Set WEBSEARCH_PROVIDER=none to disable.
package tool

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"net/http"
	"net/url"
	"regexp"
	"strings"
	"time"
)

const (
	wsDefaultResults = 5
	wsMaxResults     = 10
	wsTimeout        = 25 * time.Second
	wsUA             = "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/124.0 Safari/537.36"
)

type searchHit struct {
	Title   string
	URL     string
	Content string
}

// resolveProvider picks the active provider + key. An explicit WEBSEARCH_PROVIDER
// wins; otherwise prefer a configured API key (Tavily first), else the keyless
// DuckDuckGo scrape so search works without any configuration. "none" disables it.
func resolveProvider() (provider, key string) {
	explicit := strings.ToLower(readEnv("WEBSEARCH_PROVIDER"))
	tavily := readEnv("TAVILY_API_KEY")
	exa := readEnv("EXA_API_KEY")
	if explicit == "none" {
		return "none", ""
	}
	if explicit == "scrape" {
		return "scrape", ""
	}
	if explicit == "tavily" && tavily != "" {
		return "tavily", tavily
	}
	if explicit == "exa" && exa != "" {
		return "exa", exa
	}
	if tavily != "" {
		return "tavily", tavily
	}
	if exa != "" {
		return "exa", exa
	}
	return "scrape", ""
}

func searchTavily(ctx context.Context, query string, n int, key string) ([]searchHit, error) {
	payload, _ := json.Marshal(map[string]any{"api_key": key, "query": query, "max_results": n, "search_depth": "basic"})
	req, _ := http.NewRequestWithContext(ctx, "POST", "https://api.tavily.com/search", bytes.NewReader(payload))
	req.Header.Set("Content-Type", "application/json")
	resp, err := http.DefaultClient.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	if resp.StatusCode != 200 {
		return nil, fmt.Errorf("Tavily HTTP %d", resp.StatusCode)
	}
	var data struct {
		Results []struct {
			Title   string `json:"title"`
			URL     string `json:"url"`
			Content string `json:"content"`
		} `json:"results"`
	}
	if err := json.NewDecoder(resp.Body).Decode(&data); err != nil {
		return nil, err
	}
	var hits []searchHit
	for _, r := range data.Results {
		hits = append(hits, searchHit{Title: r.Title, URL: r.URL, Content: r.Content})
	}
	return hits, nil
}

func searchExa(ctx context.Context, query string, n int, key string) ([]searchHit, error) {
	payload, _ := json.Marshal(map[string]any{
		"query": query, "numResults": n,
		"contents": map[string]any{"text": map[string]any{"maxCharacters": 1200}},
	})
	req, _ := http.NewRequestWithContext(ctx, "POST", "https://api.exa.ai/search", bytes.NewReader(payload))
	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("x-api-key", key)
	resp, err := http.DefaultClient.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	if resp.StatusCode != 200 {
		return nil, fmt.Errorf("Exa HTTP %d", resp.StatusCode)
	}
	var data struct {
		Results []struct {
			Title string `json:"title"`
			URL   string `json:"url"`
			Text  string `json:"text"`
		} `json:"results"`
	}
	if err := json.NewDecoder(resp.Body).Decode(&data); err != nil {
		return nil, err
	}
	var hits []searchHit
	for _, r := range data.Results {
		hits = append(hits, searchHit{Title: r.Title, URL: r.URL, Content: r.Text})
	}
	return hits, nil
}

// searchScrape queries DuckDuckGo's HTML endpoint (keyless). It returns static
// HTML with far less bot-blocking than Google/Bing; still rate-limited and the
// layout can drift, so the parser is defensive and failures are treated as soft.
func searchScrape(ctx context.Context, query string, n int) ([]searchHit, error) {
	form := url.Values{"q": {query}}
	req, _ := http.NewRequestWithContext(ctx, "POST", "https://html.duckduckgo.com/html/", strings.NewReader(form.Encode()))
	req.Header.Set("User-Agent", wsUA)
	req.Header.Set("Content-Type", "application/x-www-form-urlencoded")
	req.Header.Set("Accept", "text/html")
	resp, err := http.DefaultClient.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	if resp.StatusCode != 200 {
		return nil, fmt.Errorf("DuckDuckGo HTTP %d", resp.StatusCode)
	}
	buf := new(bytes.Buffer)
	buf.ReadFrom(resp.Body)
	hits := parseDuckDuckGoHTML(buf.String())
	if len(hits) > n {
		hits = hits[:n]
	}
	return hits, nil
}

var (
	ddgLinkRe    = regexp.MustCompile(`(?is)<a[^>]*class="[^"]*result__a[^"]*"[^>]*href="([^"]+)"[^>]*>(.*?)</a>`)
	ddgSnippetRe = regexp.MustCompile(`(?is)<a[^>]*class="[^"]*result__snippet[^"]*"[^>]*>(.*?)</a>`)
	tagRe        = regexp.MustCompile(`<[^>]+>`)
	wsRe         = regexp.MustCompile(`\s+`)
	uddgRe       = regexp.MustCompile(`[?&]uddg=([^&]+)`)
)

func ddgStripTags(s string) string {
	s = tagRe.ReplaceAllString(s, "")
	r := strings.NewReplacer(
		"&nbsp;", " ", "&amp;", "&", "&lt;", "<", "&gt;", ">",
		"&quot;", `"`, "&#x27;", "'", "&#39;", "'",
	)
	s = r.Replace(s)
	return strings.TrimSpace(wsRe.ReplaceAllString(s, " "))
}

func decodeDdgURL(href string) string {
	if m := uddgRe.FindStringSubmatch(href); m != nil {
		if dec, err := url.QueryUnescape(m[1]); err == nil {
			return dec
		}
		return m[1]
	}
	if strings.HasPrefix(href, "//") {
		return "https:" + href
	}
	return href
}

// parseDuckDuckGoHTML parses the DuckDuckGo HTML result page into hits. Exported-ish
// (package-visible) so the unit test can verify it against a captured fixture.
func parseDuckDuckGoHTML(html string) []searchHit {
	var hits []searchHit
	links := ddgLinkRe.FindAllStringSubmatchIndex(html, -1)
	snips := ddgSnippetRe.FindAllStringSubmatchIndex(html, -1)
	for _, lm := range links {
		href := html[lm[2]:lm[3]]
		title := ddgStripTags(html[lm[4]:lm[5]])
		u := decodeDdgURL(href)
		if title == "" || !(strings.HasPrefix(u, "http://") || strings.HasPrefix(u, "https://")) {
			continue
		}
		content := ""
		for _, sm := range snips {
			if sm[0] > lm[0] { // first snippet appearing after this link
				content = ddgStripTags(html[sm[2]:sm[3]])
				break
			}
		}
		hits = append(hits, searchHit{Title: title, URL: u, Content: content})
	}
	return hits
}

// WebSearchTool searches the web for up-to-date information.
var WebSearchTool = Def{
	ID: "websearch",
	Description: "Search the web for up-to-date information (current events, recent docs, version-specific facts beyond your training). " +
		"Returns titles, URLs, and content snippets. Use webfetch afterward to read a specific result in full.",
	Parameters: []Param{
		{Name: "query", Type: "string", Description: "The search query", Required: true},
		{Name: "numResults", Type: "number", Description: fmt.Sprintf("Number of results (default %d, max %d)", wsDefaultResults, wsMaxResults)},
	},
	Validate: func(args map[string]any) (map[string]any, error) {
		query, err := reqString(args, "query")
		if err != nil {
			return nil, err
		}
		n, hasN, err := optNumber(args, "numResults")
		if err != nil {
			return nil, err
		}
		out := map[string]any{"query": query}
		if hasN {
			out["numResults"] = n
		}
		return out, nil
	},
	Execute: func(args map[string]any, ctx *Context) (Result, error) {
		query := args["query"].(string)
		provider, key := resolveProvider()
		if provider == "none" {
			return Result{
				Title:    query,
				Output:   "Web search is disabled (WEBSEARCH_PROVIDER=none). Proceed using your existing knowledge.",
				Metadata: map[string]any{"error": "disabled"},
			}, nil
		}
		n := wsDefaultResults
		if v, ok := args["numResults"].(float64); ok && v > 0 {
			n = int(v)
		}
		if n > wsMaxResults {
			n = wsMaxResults
		}
		cctx, cancel := context.WithTimeout(context.Background(), wsTimeout)
		defer cancel()
		var hits []searchHit
		var err error
		switch provider {
		case "tavily":
			hits, err = searchTavily(cctx, query, n, key)
		case "exa":
			hits, err = searchExa(cctx, query, n, key)
		default:
			hits, err = searchScrape(cctx, query, n)
		}
		if err != nil {
			hint := ""
			if provider == "scrape" {
				hint = " (keyless DuckDuckGo scrape — set TAVILY_API_KEY for reliable search)"
			}
			return Result{Title: query, Output: "Web search failed: " + err.Error() + hint, Metadata: map[string]any{"error": "search_failed", "provider": provider}}, nil
		}
		if len(hits) == 0 {
			return Result{Title: query, Output: "No search results found. Try a different query.", Metadata: map[string]any{"provider": provider, "results": 0}}, nil
		}
		var b strings.Builder
		fmt.Fprintf(&b, "Web search results (%s) for %q:\n\n", provider, query)
		for i, h := range hits {
			snippet := strings.Join(strings.Fields(h.Content), " ")
			if len(snippet) > 500 {
				snippet = snippet[:500]
			}
			fmt.Fprintf(&b, "%d. %s\n   %s\n   %s\n\n", i+1, h.Title, h.URL, snippet)
		}
		out, truncated, outputPath := truncateOutput(ctx, "websearch", strings.TrimRight(b.String(), "\n"), outputCap)
		return Result{Title: query, Output: out, Truncated: truncated, OutputPath: outputPath, Metadata: map[string]any{"provider": provider, "results": len(hits)}}, nil
	},
}

File: impl/ts/src/tool/websearch.ts
// websearch — search the web for up-to-date information. See docs/harness-design.md.
// Provider order: an API key (Tavily/Exa) if configured, else a keyless scrape of
// DuckDuckGo's HTML endpoint. The scrape path means search works out of the box
// without a key, at the cost of lower reliability (rate-limits, layout drift) — so
// it fails soft (empty/clear message) rather than crashing the agent loop. Set
// WEBSEARCH_PROVIDER=none to disable web search entirely.

import type { ToolDef, ToolContext, ToolResult } from "./types.js";
import { reqString, optNumber } from "./validate.js";
import { truncateOutput } from "./truncate.js";
import { readEnv } from "./env.js";

interface Args {
  query: string;
  numResults?: number;
}

const DEFAULT_RESULTS = 5;
const MAX_RESULTS = 10;
const TIMEOUT_MS = 25_000;
const UA =
  "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/124.0 Safari/537.36";

type Provider = "tavily" | "exa" | "scrape" | "none";

interface SearchHit {
  title: string;
  url: string;
  content: string;
}

// Resolve the active provider + key. An explicit WEBSEARCH_PROVIDER wins; otherwise
// use whichever API key is present (Tavily preferred), and fall back to the keyless
// DuckDuckGo scrape so search works without any configuration. "none" disables it.
function resolveProvider(): { provider: Provider; key: string } {
  const explicit = readEnv("WEBSEARCH_PROVIDER").toLowerCase();
  const tavily = readEnv("TAVILY_API_KEY");
  const exa = readEnv("EXA_API_KEY");
  if (explicit === "none") return { provider: "none", key: "" };
  if (explicit === "scrape") return { provider: "scrape", key: "" };
  if (explicit === "tavily" && tavily) return { provider: "tavily", key: tavily };
  if (explicit === "exa" && exa) return { provider: "exa", key: exa };
  // No explicit choice: prefer a configured API key, else keyless scrape.
  if (tavily) return { provider: "tavily", key: tavily };
  if (exa) return { provider: "exa", key: exa };
  return { provider: "scrape", key: "" };
}

async function searchTavily(query: string, n: number, key: string, signal: AbortSignal): Promise<SearchHit[]> {
  const resp = await fetch("https://api.tavily.com/search", {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ api_key: key, query, max_results: n, search_depth: "basic" }),
    signal,
  });
  if (!resp.ok) throw new Error(`Tavily HTTP ${resp.status}: ${(await resp.text()).slice(0, 200)}`);
  const data = (await resp.json()) as { results?: Array<{ title?: string; url?: string; content?: string }> };
  return (data.results ?? []).map((r) => ({ title: r.title ?? "", url: r.url ?? "", content: r.content ?? "" }));
}

async function searchExa(query: string, n: number, key: string, signal: AbortSignal): Promise<SearchHit[]> {
  const resp = await fetch("https://api.exa.ai/search", {
    method: "POST",
    headers: { "Content-Type": "application/json", "x-api-key": key },
    body: JSON.stringify({ query, numResults: n, contents: { text: { maxCharacters: 1200 } } }),
    signal,
  });
  if (!resp.ok) throw new Error(`Exa HTTP ${resp.status}: ${(await resp.text()).slice(0, 200)}`);
  const data = (await resp.json()) as { results?: Array<{ title?: string; url?: string; text?: string }> };
  return (data.results ?? []).map((r) => ({ title: r.title ?? "", url: r.url ?? "", content: r.text ?? "" }));
}

// Keyless search via DuckDuckGo's HTML endpoint (https://html.duckduckgo.com/html/).
// It returns static HTML with far less bot-blocking than Google/Bing, which makes it
// usable for light scraping. Still rate-limited and layout can drift, so the parser
// is defensive and the caller treats failures/empties as soft.
async function searchScrape(query: string, n: number, signal: AbortSignal): Promise<SearchHit[]> {
  const resp = await fetch("https://html.duckduckgo.com/html/", {
    method: "POST",
    headers: { "User-Agent": UA, "Content-Type": "application/x-www-form-urlencoded", Accept: "text/html" },
    body: new URLSearchParams({ q: query }).toString(),
    signal,
  });
  if (!resp.ok) throw new Error(`DuckDuckGo HTTP ${resp.status}: ${(await resp.text()).slice(0, 200)}`);
  const html = await resp.text();
  return parseDuckDuckGoHtml(html).slice(0, n);
}

// Decode the &uddg= redirect DuckDuckGo wraps result links in, and basic entities.
function decodeDdgUrl(href: string): string {
  let url = href;
  const m = href.match(/[?&]uddg=([^&]+)/);
  if (m) {
    try {
      url = decodeURIComponent(m[1]);
    } catch {
      url = m[1];
    }
  } else if (href.startsWith("//")) {
    url = "https:" + href;
  }
  return url;
}

function stripTags(s: string): string {
  return s
    .replace(/<[^>]+>/g, "")
    .replace(/&nbsp;/g, " ")
    .replace(/&amp;/g, "&")
    .replace(/&lt;/g, "<")
    .replace(/&gt;/g, ">")
    .replace(/&quot;/g, '"')
    .replace(/&#x27;|&#39;/g, "'")
    .replace(/\s+/g, " ")
    .trim();
}

// Parse the DuckDuckGo HTML result page into hits. Exported for unit testing so we
// can verify the parser against a captured fixture without hitting the network.
export function parseDuckDuckGoHtml(html: string): SearchHit[] {
  const hits: SearchHit[] = [];
  // Each result is an <a class="result__a" href="...">title</a>; the snippet is the
  // following <a class="result__snippet">...</a>. Match links first, then look for
  // the nearest snippet after each.
  const linkRe = /<a[^>]*class="[^"]*result__a[^"]*"[^>]*href="([^"]+)"[^>]*>([\s\S]*?)<\/a>/gi;
  const snippetRe = /<a[^>]*class="[^"]*result__snippet[^"]*"[^>]*>([\s\S]*?)<\/a>/gi;

  const snippets: Array<{ index: number; text: string }> = [];
  let sm: RegExpExecArray | null;
  while ((sm = snippetRe.exec(html)) !== null) {
    snippets.push({ index: sm.index, text: stripTags(sm[1]) });
  }

  let lm: RegExpExecArray | null;
  while ((lm = linkRe.exec(html)) !== null) {
    const url = decodeDdgUrl(lm[1]);
    const title = stripTags(lm[2]);
    if (!title || !/^https?:\/\//i.test(url)) continue;
    // Find the first snippet that appears after this link in the document.
    const after = snippets.find((s) => s.index > lm!.index);
    hits.push({ title, url, content: after ? after.text : "" });
  }
  return hits;
}

export const WebSearchTool: ToolDef<Args> = {
  id: "websearch",
  description:
    "Search the web for up-to-date information (current events, recent docs, version-specific facts beyond your training). Returns titles, URLs, and content snippets. Use webfetch afterward to read a specific result in full.",
  parameters: {
    query: { type: "string", description: "The search query", required: true },
    numResults: { type: "number", description: `Number of results (default ${DEFAULT_RESULTS}, max ${MAX_RESULTS})` },
  },
  validate(args) {
    const query = reqString(args, "query");
    const numResults = optNumber(args, "numResults");
    return { query, numResults };
  },
  async execute(args, ctx: ToolContext): Promise<ToolResult> {
    const { provider, key } = resolveProvider();
    if (provider === "none") {
      return {
        title: args.query,
        output:
          "Web search is disabled (WEBSEARCH_PROVIDER=none). Proceed using your existing knowledge.",
        metadata: { error: "disabled" },
      };
    }
    const n = Math.min(args.numResults && args.numResults > 0 ? Math.floor(args.numResults) : DEFAULT_RESULTS, MAX_RESULTS);
    const controller = new AbortController();
    const timer = setTimeout(() => controller.abort(), TIMEOUT_MS);
    if (ctx.signal) ctx.signal.addEventListener("abort", () => controller.abort(), { once: true });
    let hits: SearchHit[];
    try {
      hits =
        provider === "tavily"
          ? await searchTavily(args.query, n, key, controller.signal)
          : provider === "exa"
            ? await searchExa(args.query, n, key, controller.signal)
            : await searchScrape(args.query, n, controller.signal);
    } catch (e) {
      clearTimeout(timer);
      const msg = controller.signal.aborted ? "search request timed out" : String((e as Error).message ?? e);
      // Scrape is best-effort; phrase its failure as soft so the model proceeds.
      const hint = provider === "scrape" ? " (keyless DuckDuckGo scrape — set TAVILY_API_KEY for reliable search)" : "";
      return { title: args.query, output: `Web search failed: ${msg}${hint}`, metadata: { error: "search_failed", provider } };
    }
    clearTimeout(timer);

    if (hits.length === 0) {
      return { title: args.query, output: "No search results found. Try a different query.", metadata: { provider, results: 0 } };
    }
    const body = hits
      .map((h, i) => `${i + 1}. ${h.title}\n   ${h.url}\n   ${h.content.replace(/\s+/g, " ").trim().slice(0, 500)}`)
      .join("\n\n");
    const shaped = truncateOutput(ctx, "websearch", `Web search results (${provider}) for "${args.query}":\n\n${body}`);
    return {
      title: args.query,
      output: shaped.output,
      truncated: shaped.truncated,
      outputPath: shaped.outputPath,
      metadata: { provider, results: hits.length },
    };
  },
};
